### ISSUE TITLE
Scala Enumeration: schema drops ordering and annotations

### ISSUE THREAD

**Issue by klass-ivan** [sksamuel/avro4s#299]
```scala
package example

@AvroDoc("just numbers")
@AvroProp("deprecated", "true")
object Example extends Enumeration {
  val one, two, three, four, five = Value
}
```
```scala
AvroSchema[example.Example.Value]
```
```js 
{
  "type" : "enum",
  "name" : "Example",
  "namespace" : "example",
  "symbols" : [ "five", "four", "one", "three", "two" ]
}
```
The symbols ordering is lost, they are ordered alphabetically (see [scalaEnumSchemaFor](https://github.com/sksamuel/avro4s/blob/ad4b8936098c03852c154022f66bdbb4090b8bc7/avro4s-macros/src/main/scala/com/sksamuel/avro4s/SchemaFor.scala#L201) ) It does not directly affect avro encoding&decoding, but may be very confusing and misleading for use-cases when avro is used as interchange format across different languages, where schema generated from scala may be used to generate definitions in other languages.
Also, all annotations except name & namespace are dropped.

Here is a quick no-macros workaround I came up to:
```scala
trait AvroEnum extends Enumeration {
  enum =>

  import scala.tools.reflect.ToolBox
  import scala.reflect.runtime.currentMirror
  import scala.reflect.runtime.universe._

  implicit def schemaFor[E <: Value](implicit tag: TypeTag[E]): SchemaFor[E] = new SchemaFor[E] {
    val typeRef = tag.tpe match {
      case t @ TypeRef(_, _, _) => t
    }

    val annos = typeRef.pre.typeSymbol.annotations.map { a =>
      val name = a.tree.tpe.typeSymbol.fullName
      val tb   = currentMirror.mkToolBox()
      val args = tb.compile(tb.parse(a.toString)).apply() match {
        case c: AvroFieldReflection => c.getAllFields.map { case (k, v) => (k, v.toString) }
        case _                      => Map.empty[String, String]
      }
      Anno(name, args)
    }

    val extractor = new AnnotationExtractors(annos)
    override def schema: Schema =
      extractor.props
        .foldLeft(
          SchemaBuilder
            .enumeration(
              extractor.name.getOrElse(
                typeRef.pre.typeSymbol.name.decodedName.toString
              )
            )
            .doc(
              extractor.doc.orNull
            )
            .namespace(
              extractor.namespace.getOrElse(
                typeRef.pre.typeSymbol.owner.fullName
              )
            )
        )(
          (schemaBuilder, kv) => schemaBuilder.prop(kv._1, kv._2)
        )
        .symbols(
          enum.values.toList.map(_.toString): _* // keep symbol ordering
        )
  }
}
```

Using it instead of `Enumeration` brings proper schema derivation

```scala

@AvroDoc("just numbers")
@AvroProp("deprecated", "true")
object Example extends AvroEnum {
  val one, two, three, four, five = Value
}
```

```AvroSchema[example.Example.Value]```

```js
{
  "type" : "enum",
  "name" : "Example",
  "namespace" : "example",
  "doc" : "just numbers",
  "symbols" : [ "one", "two", "three", "four", "five" ],
  "deprecated" : "true"
}
```

It would be great to have more "native" implementation with annotations and symbols ordering, without having to use such mixins. 
I'm not so good in macros yet to have a PR, but hope these code snippets may help. 
Thank you!

**Comment by klass-ivan** [comment]
Realized that more correctly annotation type should be not on the enumeration object but on the `Value` type. But ordering is still a problem

**Comment by klass-ivan** [comment]
Looks like the order still could be preserved when using shapeless witness and no special trait inheritance:
```scala
implicit def scalaEnumSchemaFor[E <: Enumeration](
  implicit tag: TypeTag[E#Value],
  enum: Witness.Aux[E]
): SchemaFor[E#Value] = new SchemaFor[E#Value] {
  val typeRef = tag.tpe match {
    case t @ TypeRef(_, _, _) => t
  }
  val annos = typeRef.pre.typeSymbol.annotations.map { a =>
    val name = a.tree.tpe.typeSymbol.fullName
    val tb   = currentMirror.mkToolBox()
    val args = tb.compile(tb.parse(a.toString)).apply() match {
      case c: AvroFieldReflection => c.getAllFields.map { case (k, v) => (k, v.toString) }
      case _                      => Map.empty[String, String]
    }
    Anno(name, args)
  }

  val extractor = new AnnotationExtractors(annos)

  override def schema: Schema =
    extractor.props
      .foldLeft(
        SchemaBuilder
          .enumeration(
            extractor.name.getOrElse(
              typeRef.pre.typeSymbol.name.decodedName.toString
            )
          )
          .doc(
            extractor.doc.orNull
          )
          .namespace(
            extractor.namespace.getOrElse(
              typeRef.pre.typeSymbol.owner.fullName
            )
          )
      )(
        (schemaBuilder, kv) => schemaBuilder.prop(kv._1, kv._2)
      )
      .symbols(
        enum.value.values.toList.map(_.toString): _* // keep symbol ordering
      )
}
```

**Comment by sksamuel** [comment]
Just dropping the .sorted from the scalaEnumSchemaFor should be sufficient to fix this right ?

**Comment by klass-ivan** [comment]
@sksamuel If there is a guarantee that `typeRef.pre.typeSymbol.typeSignature.members` are definition-sorted, than right, just removing `.sorted` is ok

**Comment by sksamuel** [comment]
Can you try the 3.0.x snapshot from 
https://oss.sonatype.org/content/repositories/snapshots/com/sksamuel/avro4s/avro4s-core_2.12/
This should be fixed and is available for 2.11 and 2.22

**Comment by klass-ivan** [comment]
@sksamuel I've checked avro4s 3.0 source and I see it uses filtered and re-versed `typeSymbol.typeSignature.members`  but scala API says:

     ...
     *  Members in the returned scope might appear in arbitrary order.
     *  Use `declarations.sorted` to get an ordered list of members.
     */
    def members: MemberScope

Which means we can't rely on this

**Comment by sksamuel** [comment]
Isn't .sorted just what we had in the original version!?

**Comment by sksamuel** [comment]
I've added .sorted back in, and it's in reverse order now.

**Comment by sksamuel** [comment]
I had an extra reverse in there. So back to .sorted. If you're happy we can
close this.

**Comment by klass-ivan** [comment]
@sksamuel in 2.0.4 `.sorted` is on list of strings, now we have `.sorted` on list of member symbols, that means definition order, which I hope works. Will check now

**Comment by klass-ivan** [comment]
@sksamuel works on snapshot - thank you very much for the fix, closing & waiting for stable 3.0 release.
Extra thumb up for naming strategy in SchemaFor :+1:
